tasmota/xsns_96_bl6523.ino

```diff
@@ -0,0 +1,353 @@
+    */
+   powf = 0.0f;
+   powf_word = Bl6523.powf & 0x7fffff; //Extract the 23 bits
+   for (int i = 0; i < 23; i++){ // Accumulate powf from 23 bits
+    powf += ((powf_word >> (22-i)) * pow(2,(0-(i+1)))); 
+    powf_word = powf_word & (0x7fffff >> (1+i));
+   }
+   powf = (Bl6523.powf >> 23) ? (0.0f - (~powf)) : powf; // Negate if sign bit(24) is set
+   
+   watthr = (float)Bl6523.watthr / BL6523_DIV_WATTHR;
+
+   ext_snprintf_P(amps_str, sizeof(amps_str), PSTR("%3_f"), &amps);
+   ext_snprintf_P(volts_str, sizeof(volts_str), PSTR("%2_f"), &volts);
+   ext_snprintf_P(freq_str, sizeof(freq_str), PSTR("%2_f"), &freq);
+   ext_snprintf_P(watts_str, sizeof(watts_str), PSTR("%3_f"), &watts);
+   ext_snprintf_P(powf_str, sizeof(powf_str), PSTR("%2_f"), &powf);
+   ext_snprintf_P(watthr_str, sizeof(watthr_str), PSTR("%3_f"), &watthr);
+
+    if (json)
+    {
+      ResponseAppend_P(PSTR(",\"BL6523\":{"));
+      ResponseAppend_P(PSTR("\"Amps\":%s,"), amps_str);
+      ResponseAppend_P(PSTR("\"Volts\":%s,"), volts_str);
+      ResponseAppend_P(PSTR("\"Freq\":%s,"), freq_str);
+      ResponseAppend_P(PSTR("\"Watts\":%s,"), watts_str);
+      ResponseAppend_P(PSTR("\"Powf\":%s,"), powf_str);
+      ResponseAppend_P(PSTR("\"WattHr\":%s"), watthr_str);
+      ResponseJsonEnd();
+#ifdef USE_DOMOTICZ
+      if (0 == TasmotaGlobal.tele_period)
+      {
+        DomoticzSensor(DZ_CURRENT, amps_str);    // Amps
+        DomoticzSensor(DZ_VOLTAGE, volts_str); // Voltage
+        DomoticzSensor(DZ_COUNT, freq_str);  // Frequency
+        DomoticzSensor(DZ_ILLUMINANCE, watts_str);    // Watts
+        DomoticzSensor(DZ_P1_SMART_METER, powf_str); // Power Factor
+        DomoticzSensor(DZ_POWER_ENERGY, watthr_str);  // WattHour
+      }
+#endif // USE_DOMOTICZ
+#ifdef USE_WEBSERVER
+    }
+    else
+    {
+      WSContentSend_PD(HTTP_BL6523_SNM);
+      WSContentSend_PD(HTTP_BL6523_SNS, PSTR("Amps:"), amps_str);
+      WSContentSend_PD(HTTP_BL6523_SNS, PSTR("Volts:"), volts_str);
+      WSContentSend_PD(HTTP_BL6523_SNS, PSTR("Freq:"), freq_str);
+      WSContentSend_PD(HTTP_BL6523_SNS, PSTR("Watts:"), watts_str);
+      WSContentSend_PD(HTTP_BL6523_SNS, PSTR("PowF:"), powf_str);
+      WSContentSend_PD(HTTP_BL6523_SNS, PSTR("WattHr:"), watthr_str);
+#endif // USE_WEBSERVER
+    }
+  }
+}
+
+/*********************************************************************************************\
+ * Interface
+\*********************************************************************************************/
+
+bool Xsns96(uint8_t function)
+{
+  bool result = false;
+
+  if (Bl6523.type)
+  {
+    switch (function)
+    {
+    case FUNC_EVERY_250_MSECOND:
+      Bl6523Update();
+      break;
+    case FUNC_JSON_APPEND:
+      Bl6523Show(1);
+      break;
+#ifdef USE_WEBSERVER
+    case FUNC_WEB_SENSOR:
+      Bl6523Show(0);
+      break;
+#endif // USE_WEBSERVER
+    case FUNC_INIT:
+      Bl6523Init();
+      break;
+    }
+  }
+  return result;
+}
```

Comment:
This structure is not correct
It should be:
```C++
if ( FUNC_INIT == function )
{
   BL6523Init();
}
else if ( Bl6523.type ) {  // Bl6523.type must be defaulting to 0 and set to 1 in INIT is the driver is enabled only
    switch( funcstion ) {
    ....
}
```

Reply:
I changed as per comment. But I can see that many places in existing code uses the previous structure as well.